Added documentation for workspaces in O3 (#11)
denniskigen/o3-docs: pages/docs/workspaces-in-o3/registering-workspaces.mdx

# Registering workspaces

## Workspace Properties

1. **name**: string - Unique identifier for the workspace
2. **title**: string - Title shown in the workspace header
3. **load**: Function - A lifecycle function that loads the workspace component (see examples)
4. **type**: string - type of the workspace (Default = form)
5. **canHide**: boolean - Can the workspace be collapsed/hidden (Default = false)
6. **canMaximize**: boolean - Can the workspace be maximized to the full screen (Default = false)
7. **width**: 'narrow' | 'wider' - Whether the workspace should have a width of 420px (narrow) or 520px (wider; especially for forms with navigation)
8. **preferredWindowSize**: 'normal' | 'maximized' - Whether to open the workspace in normal view or in maximized view (Default = 'normal')

> 💡 Please note that `canHide` property should be `true` for workspaces with a `type` that corresponds to a siderail nav button.

> Registering a new workspace can be simply done in either of the 2 ways:
> 
> ## Registering workspace in `index.ts` (Recommended)
> 
> A new workspace can be registered in the `index.ts` file, using the `registerWorkspace` function from the `@openmrs/esm-patient-common-lib`

Reviewer: ```suggestion
## How to Register a Workspace

A new workspace can be registered in the `index.ts` file, using the `registerWorkspace` function from `@openmrs/esm-patient-common-lib`.
```

```ts copy
// index.ts
import { registerWorkspace } from '@openmrs/esm-patient-common-lib';

registerWorkspace({

>   name: string,
>   title: string,
>   load: getAsyncLifecycle(() => import(<path-to-component>), options),
>   type: string,
>   canHide?: boolean = false,
>   canMaximize?: boolean = false,
>   width?: 'narrow' | 'wider' = 'narrow',
>   preferredWindowSize?: 'normal' | 'maximized' = 'normal'

Reviewer: ```suggestion
  name: "gastrodynamics-form,
  title: translateFrom("@myorg/esm-gastrodynamics-app", "gastrodynamicsTitle", "Gastrodynamics"),
  load: getAsyncLifecycle(() => import("./gastrodynamics.workspace"), options),
  type: "other-form",
  canHide: false,
  canMaximize: true,
  width: 'narrow',
  preferredWindowSize: 'maximized'
```

> 
> registerWorkspace({
>   name: string,
>   title: string,
>   load: getAsyncLifecycle(() => import(<path-to-component>), options),
>   type: string,
>   canHide?: boolean = false,
>   canMaximize?: boolean = false,
>   width?: 'narrow' | 'wider' = 'narrow',
>   preferredWindowSize?: 'normal' | 'maximized' = 'normal'
> });

Reviewer: ```suggestion

export function startupApp() {
  registerWorkspace({
    name: "gastrodynamics-form,
    title: translateFrom("@myorg/esm-gastrodynamics-app", "gastrodynamicsTitle", "Gastrodynamics"),
    load: getAsyncLifecycle(() => import("./gastrodynamics.workspace"), options),
    type: "other-form",
    canHide: false,
    canMaximize: true,
    width: 'narrow',
    preferredWindowSize: 'maximized'
  });
}
```

> 
> ```

Reviewer: ```suggestion

```

You will have to create the file `./gastrodynamics.workspace.tsx` with your workspace component as a default export.
```

# Registering workspaces

## Workspace Properties

Workspace have a few properties defining the structure and behaviour of handling the workspace window.

1. **name**: string - Unique identifier for the workspace.
2. **title**: string - Title shown in the workspace header.
3. **load**: Function - path to the component to be rendered.
4. **type**: string - type of the workspace (Default = form)
5. **canHide**: boolean - can the workspace be collapsed/ hidden (Default = false)
6. **canMaximize**: boolean - can the workspace be maximized to the full screen(Default = false)

> 7. **width**: 'narrow' | 'wider' - Whether the workspace should have a width of 420px (narrow) or 520px (wide, especially for forms with sub forms (sub-sections)) width.

Reviewer 2: ```suggestion
7. **width**: 'narrow' | 'wider' - Whether the workspace should have a width of 420px (narrow) or 520px (wide, specifically for forms with navigation) width.
```

8. **preferredWindowSize**: 'normal' | 'maximized' - Whether to open the workspace in normal view or in maximized view. (Default = 'normal')

> 💡 Please note that `canHide` property should be allowed for workspaces having their corresponding siderail nav button

Registering a new workspace can be simply done in either of the 2 ways:

## Registering workspace in `index.ts` (Recommended)

A new workspace can be registered in the `index.ts` file, using the `registerWorkspace` function from the `@openmrs/esm-patient-common-lib`

```ts copy
// index.ts
import { registerWorkspace } from '@openmrs/esm-patient-common-lib';

registerWorkspace({
  name: string,
  title: string,
  load: getAsyncLifecycle(() => import(<path-to-component>), options),
  type: string,
  canHide?: boolean = false,
  canMaximize?: boolean = false,
  width?: 'narrow' | 'wider' = 'narrow',
  preferredWindowSize?: 'normal' | 'maximized' = 'normal'
});

```

> ## Registering workspace as an extension in `routes.json`

Reviewer 2: This should probably note that it's not the favored way of doing this. Although, and I've been saying this for years now, we should make the workspace into a site-wide concept.
